api/types/container/network.go

```diff
@@ -0,0 +1,308 @@
+//
+// The zero port range value, i.e. PortRange{}, is invalid; use [ParsePortRange] to create a valid PortRange value.
+type PortRange struct {
+	start uint16
+	end   uint16
+	proto unique.Handle[PortProto]
+}
+
+// ParsePortRange parses s as a [PortRange].
+//
+// It normalizes the provided protocol such that "80-90/tcp", "80-90/TCP", and "80-90/tCp" are equivalent.
+// If a port number range is provided, but no protocol, the default ("tcp") protocol is returned.
+func ParsePortRange(s string) (PortRange, error) {
+	if s == "" {
+		return PortRange{}, errors.New("value is empty")
+	}
+
+	portRange, proto, _ := strings.Cut(s, "/")
+
+	start, end, ok := strings.Cut(portRange, "-")
+	startVal, err := parsePortNumber(start)
+	if err != nil {
+		return PortRange{}, fmt.Errorf("invalid start port '%s': %w", start, err)
+	}
+
+	portProto := normalizePortProto(proto)
+
+	if !ok || start == end {
+		return PortRange{start: startVal, end: startVal, proto: unique.Make(portProto)}, nil
+	}
+
+	endVal, err := parsePortNumber(end)
+	if err != nil {
+		return PortRange{}, fmt.Errorf("invalid end port '%s': %w", end, err)
+	}
+	if endVal < startVal {
+		return PortRange{}, errors.New("invalid port range: " + s)
+	}
+	return PortRange{start: startVal, end: endVal, proto: unique.Make(portProto)}, nil
+}
+
+// MustParsePortRange calls [ParsePortRange](s) and panics on error.
+// It is intended for use in tests with hard-coded strings.
+func MustParsePortRange(s string) PortRange {
+	pr, err := ParsePortRange(s)
+	if err != nil {
+		panic(err)
+	}
+	return pr
+}
+
+// PortRangeFrom returns a [PortRange] with the given start and end port numbers and protocol.
+//
+// If end < start, then PortRangeFrom returns an error.
+func PortRangeFrom(start, end uint16, proto PortProto) (PortRange, error) {
+	if end < start {
+		return PortRange{}, errors.New("invalid port range")
+	}
+	if proto == "" {
+		return PortRange{}, nil
+	}
+	normalized := normalizePortProto(string(proto))
+	return PortRange{start: start, end: end, proto: unique.Make(normalized)}, nil
+}
+
+func (pr PortRange) Start() uint16 {
+	return pr.start
+}
+
+func (pr PortRange) End() uint16 {
+	return pr.end
+}
+
+func (pr PortRange) Proto() PortProto {
+	return pr.proto.Value()
+}
+
+// IsZero reports whether the [PortRange] is the zero value.
+func (pr PortRange) IsZero() bool {
+	return pr.proto == protoZero
+}
+
+// IsValid reports whether the [PortRange] is an initialized valid port range (not the zero value).
+func (pr PortRange) IsValid() bool {
+	return pr.proto != protoZero
+}
+
+func (pr PortRange) String() string {
+	switch pr.proto {
+	case protoZero:
+		return "invalid port range"
+	default:
+		return string(pr.appendText(nil))
+	}
+}
+
+func (pr PortRange) AppendText(b []byte) ([]byte, error) {
+	return pr.appendText(b), nil
+}
+
+func (pr PortRange) appendText(b []byte) []byte {
+	if pr.IsZero() {
+		return b
+	}
+	return fmt.Appendf(b, "%d-%d/%s", pr.start, pr.end, pr.proto.Value())
```

Comment:
I wonder if we should marshal a range to the shortest string that parses back to the original value.
```suggestion
	if pr.start == pr.end {
		return fmt.Appendf(b, "%d/%s", pr.start, pr.proto.Value())
	}
	return fmt.Appendf(b, "%d-%d/%s", pr.start, pr.end, pr.proto.Value())
```

If we do that, the canonical textual representation of a single-port range would also be accepted by `ParsePort`. I don't think that would be an issue in practice as clients instantiate API structs from type definitions we provide. There would not be an opportunity to incorrectly use a `PortRange` in place of a `Port` or vice versa.

Reply:
👍🏼 makes sense.